src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalTabPresenter.java

```diff
@@ -43,62 +47,76 @@
        * Ensure terminal pane is visible.
        */
       void activateTerminal();
-      
-      /**
-       *  Ensure terminal pane has at least one session loaded.
-       */
-      void ensureTerminal();
-      
+
       /**
        * Create a new terminal session.
        */
       void createTerminal();
-      
+
       /**
        * Terminate current terminal.
        */
       void terminateCurrentTerminal();
-      
+
       /**
        * Attach a list of server-side terminals to the pane.
        * @param event list of terminals on server
        */
       void repopulateTerminals(ArrayList<ConsoleProcessInfo> procList);
+
+      /**
+       * @return Number of busy terminals (a terminal where the shell has
+       * child processes is considered busy).
+       */
+      int busyTerminalCount();
+
+      /**
+       * Terminate all terminals, whether busy or not. This kills any server-side
+       * process and removes it from the list of known processes. This should
+       * only be invoked when the terminal tab itself is being unloaded.
+       */
+      void terminateAllTerminals();
    }
 
    @Inject
    public TerminalTabPresenter(final Display view,
                                final Session session,
+                               GlobalDisplay globalDisplay,
+                               UIPrefs uiPrefs,
                                Provider<ConsoleProcessFactory> pConsoleProcessFactory)
    {
       super(view);
       view_ = view;
       session_ = session;
+      globalDisplay_ = globalDisplay;
+      uiPrefs_ = uiPrefs;
       pConsoleProcessFactory_ = pConsoleProcessFactory;
    }
-  
+
    @Handler
    public void onActivateTerminal()
    {
+      uiPrefs_.showTerminalTab().setGlobalValue(true);
```

Comment:
Update after chatting: I understand how this works better now. Let's stick with what you've got and see how it feels.